File: Runtime/InitialConditionsDistributionMap.vb
'************************************************************************************
' ST-Sim: A .NET class library for developing state-and-transition simulation models.
'
' Copyright © 2007-2017 Apex Resource Management Solution Ltd. (ApexRMS). All rights reserved.
'
'************************************************************************************

Imports SyncroSim.Common

Class InitialConditionsDistributionMap

    Private m_HasItems As Boolean
    Private m_Map As New SortedKeyMap1(Of InitialConditionsDistributionCollection)(SearchMode.ExactPrev)

    Public Sub New(ByVal icd As InitialConditionsDistributionCollection)

        For Each t As InitialConditionsDistribution In icd
            Me.AddICD(t)
        Next

    End Sub

    Private Sub AddICD(ByVal order As InitialConditionsDistribution)

        Dim l As InitialConditionsDistributionCollection = Me.m_Map.GetItemExact(order.Iteration)

        If (l Is Nothing) Then

            l = New InitialConditionsDistributionCollection()
            Me.m_Map.AddItem(order.Iteration, l)

        End If

        l.Add(order)

        Me.m_HasItems = True

    End Sub

    Public Function GetICDs(ByVal iteration As Nullable(Of Integer)) As InitialConditionsDistributionCollection

        If (Not Me.m_HasItems) Then
            Return Nothing
        End If

        Dim l As InitialConditionsDistributionCollection = Me.m_Map.GetItem(iteration)

        If (l Is Nothing) Then
            Return Nothing
        End If

        Debug.Assert(l.Count > 0)
        Return l

    End Function

End Class

File: src/Runtime/InitialConditionsDistributionMap.cs
// stsim: A SyncroSim Package for developing state-and-transition simulation models using ST-Sim.
// Copyright © 2007-2021 Apex Resource Management Solutions Ltd. (ApexRMS). All rights reserved.

using SyncroSim.Common;
using System.Diagnostics;

namespace SyncroSim.STSim
{
    internal class InitialConditionsDistributionMap
    {
        private bool m_HasItems;
        private SortedKeyMap1<InitialConditionsDistributionCollection> m_Map = 
            new SortedKeyMap1<InitialConditionsDistributionCollection>(SearchMode.ExactPrev);

        public InitialConditionsDistributionMap(InitialConditionsDistributionCollection icd)
        {
            foreach (InitialConditionsDistribution t in icd)
            {
                this.AddICD(t);
            }
        }

        private void AddICD(InitialConditionsDistribution order)
        {
            InitialConditionsDistributionCollection l = this.m_Map.GetItemExact(order.Iteration);

            if (l == null)
            {
                l = new InitialConditionsDistributionCollection();
                this.m_Map.AddItem(order.Iteration, l);
            }

            l.Add(order);

            this.m_HasItems = true;
        }

        public InitialConditionsDistributionCollection GetICDs(int? iteration)
        {
            if (!this.m_HasItems)
            {
                return null;
            }

            InitialConditionsDistributionCollection l = this.m_Map.GetItem(iteration);

            if (l == null)
            {
                return null;
            }

            Debug.Assert(l.Count > 0);
            return l;
        }
    }
}
